types/observer-mixin.d.ts

```diff
@@ -5,7 +5,11 @@
 
 import {Callback, PromiseOrVoid} from './common';
 
-export type Listener = (ctx: object, next: (err?: any) => void) => void;
+export interface OperationHookContext {
+  [property: string]: any;
```

Comment:
For longer term, I'd like us to provide strongly typed context types based on the hook name. See this doc page for more details about context objects provided by different hooks: https://loopback.io/doc/en/lb3/Operation-hooks.html#hooks

For example:

```ts
export interface AccessHookContext<T extends typeof PersistedModel> {
  Model: T;
  query: Filter<T>;
}

export type Listener<Ctx> = (ctx: Ctx, next: (err?: any) => void) => void;

class PersistedModel {
// ...
static observe(operation: 'access', handler: Listener<AccessHookContext>): void;
static observe(operation: 'before save', handler: Listener<BeforeSaveHookContext>): void;
// etc.
}
```

I think it's ok to leave that as a long term goal, out of scope of this pull request.

Having said that, I believe that all operation hooks are always providing `Model` property. Let's include it in the type definition please.

```ts
export interface OperationHookContext<T extends typeof PersistedModel> {
  /**
   * The constructor of the model that triggered the operation.
   */
  Model: T;

  /**
   * Additional context properties, not typed yet.
   * See https://loopback.io/doc/en/lb3/Operation-hooks.html#hooks
   */
  [property: string]: any;
}
```

Reply:
👍

Reply:
Created an issue to note the suggestion and follow through - https://github.com/strongloop/loopback-datasource-juggler/issues/1822.

Reply:
@bajtos @hacksparrow, would it be a `tech-debt` for Q2? or much longer term?

Reply:
Wouldn't call it a tech debt, exactly; more like something very good to have.

Reply:
Perhaps `developer-experience` + `2020Q2`?